components/abstractions/src/main/java/com/microsoft/kiota/authentication/ApiKeyAuthenticationProvider.java

```diff
@@ -0,0 +1,100 @@
+
+import io.opentelemetry.api.GlobalOpenTelemetry;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+
+/** This authentication provider adds an API key to the request as a query parameter or header. */
+public class ApiKeyAuthenticationProvider implements AuthenticationProvider {
+    private final ApiKeyLocation location;
+    @Nonnull
+    private final String paramName;
+    @Nonnull
+    private final String apiKey;
+    private final AllowedHostsValidator validator;
+    /**
+     * Creates a new instance of ApiKeyAuthenticationProvider.
+     * @param apiKey The API key to use.
+     * @param paramName The name of the query parameter or header to use.
+     * @param location The location of the API key.
+     * @param validHosts The list of allowed hosts for which to add the API key.
+     */
+    public ApiKeyAuthenticationProvider(@Nonnull final String apiKey, @Nonnull final String paramName, @Nonnull final ApiKeyLocation location, @Nonnull final String ...validHosts) {
+        Objects.requireNonNull(apiKey);
+        Objects.requireNonNull(paramName);
+        Objects.requireNonNull(location);
+        if (apiKey.isEmpty()) {
+            throw new IllegalArgumentException("apiKey cannot be empty");
+        }
+        if (paramName.isEmpty()) {
+            throw new IllegalArgumentException("paramName cannot be empty");
+        }
+        this.apiKey = apiKey;
+        this.paramName = paramName;
+        this.location = location;
+        this.validator = new AllowedHostsValidator(validHosts);
+    }
+    private final static String parentSpanKey = "parent-span";
+    /** {@inheritDoc} */
+    @Override
+    @Nonnull
+    public CompletableFuture<Void> authenticateRequest(@Nonnull final RequestInformation request, @Nullable final Map<String, Object> additionalAuthenticationContext) {
+        Objects.requireNonNull(request);
+        final CompletableFuture<Void> resultFuture = new CompletableFuture<>();
+        Span span;
+        if(additionalAuthenticationContext != null && additionalAuthenticationContext.containsKey(parentSpanKey) && additionalAuthenticationContext.get(parentSpanKey) instanceof Span) {
+            final Span parentSpan = (Span) additionalAuthenticationContext.get(parentSpanKey);
+            span = GlobalOpenTelemetry.getTracer("com.microsoft.kiota").spanBuilder("authenticateRequest").setParent(Context.current().with(parentSpan)).startSpan();
+        } else {
+            span = GlobalOpenTelemetry.getTracer("com.microsoft.kiota").spanBuilder("authenticateRequest").startSpan();
+        }
+        try(final Scope scope = span.makeCurrent()) {
+            final URI uri = request.getUri();
+            if(uri == null || !validator.isUrlHostValid(uri)) {
+                span.setAttribute("com.microsoft.kiota.authentication.is_url_valid", false);
+                return CompletableFuture.completedFuture(null);
+            }
```

Comment:
we're doing the same thing in the azure identity provider. If the host is invalid, we No-OP, and don't throw an exception. That's because of cases where we get a redirect to a different domain which is not in the valid hosts (large file upload does that for Microsoft Graph). These requests shouldn't get authenticated, but it's an expectation rather than an exception.

As for the "null" value, it's just how generic types are weird in Java and CompletableFuture<Void> can only be created from passing null as the value....